README.md
# HTTP accept header

A very basic & naive implementation of the HTTP Accept header. It uses [http](https://crates.io/crates/http) crate and [mime](https://crates.io/crates/mime) crate to parse the accept header. Basic data structure:

```rust
#[derive(Debug, Clone, PartialEq)]
pub struct Accept {
    pub wildcard: Option<MediaType>,
    pub types: Vec<MediaType>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MediaType {
    pub mime: Mime,
    pub weight: Option<f32>,
}
```

Usage:

```rust
// parse accept header
let accept: Accept = "application/json, text/html;q=0.9, text/plain;q=0.8, */*;q=0.7"
    .parse()
    .unwrap();

// prepare a list of supported media types
let available = vec![
    Mime::from_str("text/html").unwrap(),
    Mime::from_str("application/json").unwrap(),
];

// content negotiation
let negotiated = accept.negotiate(&available).unwrap();

// "application/json" shall be chosen since it is available and has the highest weight
assert_eq!(negotiated, Mime::from_str("application/json").unwrap());
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.2.3] - 2023-02-14

[1c31269](1c312693bd373d75acec6bb9598de70d052f95a5)...[f993304](f9933043fabf8d9aa9e3f69b62f6437ece27ef21)

### Features

- Change negotiate interface to use &[Mime] ([f993304](f9933043fabf8d9aa9e3f69b62f6437ece27ef21) - 2023-02-14 by Tyr Chen)

## [0.2.1] - 2022-11-27

[3d200b6](3d200b681cac791c6380b3c56318bae11da29644)...[f782be8](f782be8c55ceb425c3a3e4686923fdf24ef35b6b)

### Miscellaneous Tasks

- Change interface from vec to hashset for better performance. Also select text/html by default ([f782be8](f782be8c55ceb425c3a3e4686923fdf24ef35b6b) - 2022-11-27 by Tyr Chen)

## [0.2.0] - 2022-11-15

[1bdee46](1bdee46f40554c068e35b3f9677b7952979f5182)...[3d200b6](3d200b681cac791c6380b3c56318bae11da29644)

### Bug Fixes

- Fix weight issue based on explanation from: [MDN](https://developer.mozilla.org/en-US/docs/Glossary/Quality_values) ([3d200b6](3d200b681cac791c6380b3c56318bae11da29644) - 2022-11-15 by Tyr Chen)

## [0.1.2] - 2022-11-14

[7fa4d3c](7fa4d3cc7eacd592057a1a4ff1ef13542f3c9cc7)...[1bdee46](1bdee46f40554c068e35b3f9677b7952979f5182)

### Features

- Support to_string for MediaType and Accept. ([1bdee46](1bdee46f40554c068e35b3f9677b7952979f5182) - 2022-11-14 by Tyr Chen)

## [0.1.1] - 2022-11-14

[b578abd](b578abdee31d425667319b1af0358c16fa8fbde7)...[7fa4d3c](7fa4d3cc7eacd592057a1a4ff1ef13542f3c9cc7)

### Features

- Support From<Mime> for Accept type ([7fa4d3c](7fa4d3cc7eacd592057a1a4ff1ef13542f3c9cc7) - 2022-11-14 by Tyr Chen)

## [0.1.0] - 2022-11-14

### Features

- Support basic functionality of parsing accept header and handle content negotiation ([75defe1](75defe18db38f97135b20ab928c4a9c4cee30579) - 2022-11-14 by Tyr Chen)

<!-- generated by git-cliff -->


src/lib.rs
#![warn(
    clippy::all,
    clippy::dbg_macro,
    clippy::todo,
    clippy::empty_enum,
    clippy::enum_glob_use,
    clippy::mem_forget,
    clippy::unused_self,
    clippy::filter_map_next,
    clippy::needless_continue,
    clippy::needless_borrow,
    clippy::match_wildcard_for_single_variants,
    clippy::if_let_mutex,
    clippy::mismatched_target_os,
    clippy::await_holding_lock,
    clippy::match_on_vec_items,
    clippy::imprecise_flops,
    clippy::suboptimal_flops,
    clippy::lossy_float_literal,
    clippy::rest_pat_in_fully_bound_structs,
    clippy::fn_params_excessive_bools,
    clippy::exit,
    clippy::inefficient_to_string,
    clippy::linkedlist,
    clippy::macro_use_imports,
    clippy::option_option,
    clippy::verbose_file_reads,
    clippy::unnested_or_patterns,
    clippy::str_to_string,
    rust_2018_idioms,
    future_incompatible,
    nonstandard_style,
    missing_debug_implementations,
    // missing_docs
)]
// #![deny(unreachable_pub, private_in_public)]
#![allow(elided_lifetimes_in_paths, clippy::type_complexity)]
#![forbid(unsafe_code)]
#![cfg_attr(docsrs, feature(doc_auto_cfg, doc_cfg))]
#![cfg_attr(test, allow(clippy::float_cmp))]

mod accept;
mod media_type;

pub mod error;

use mime::Mime;

#[derive(Debug, Clone, PartialEq)]
pub struct Accept {
    pub wildcard: Option<MediaType>,
    pub types: Vec<MediaType>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MediaType {
    pub mime: Mime,
    pub weight: Option<f32>,
}


src/media_type.rs
use mime::Mime;
use snafu::{ensure, ResultExt};

use crate::{error::*, MediaType};
use std::{cmp::Ordering, fmt, str::FromStr};

impl FromStr for MediaType {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut parts = s.split(';');
        let v = parts.next().unwrap().trim();
        let mime = v.parse().context(MediaTypeSnafu { value: v })?;
        let Some(v) = parts
            .next()
            .map(|s| s.trim())
            .and_then(|s| s.strip_prefix("q=")) else {
            return Ok(MediaType { mime, weight: None });
            };

        let weight: f32 = v.trim().parse().context(ParseWeightSnafu { value: v })?;
        ensure!(
            (0.0..=1.0).contains(&weight),
            WeightRangeSnafu { value: weight }
        );

        Ok(MediaType {
            mime,
            weight: Some(weight),
        })
    }
}

impl From<Mime> for MediaType {
    fn from(mime: Mime) -> Self {
        Self { mime, weight: None }
    }
}

impl From<MediaType> for Mime {
    fn from(media_type: MediaType) -> Self {
        media_type.mime
    }
}

impl PartialEq<Mime> for MediaType {
    fn eq(&self, other: &Mime) -> bool {
        self.mime == *other
    }
}

impl PartialEq<Mime> for &MediaType {
    fn eq(&self, other: &Mime) -> bool {
        self.mime == *other
    }
}

impl PartialOrd for MediaType {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        let aweight = self.weight.unwrap_or(1.0);
        let bweight = other.weight.unwrap_or(1.0);
        match aweight.partial_cmp(&bweight) {
            Some(Ordering::Equal) => match (self.mime.type_(), other.mime.type_()) {
                (mime::STAR, mime::STAR) => match (self.mime.subtype(), other.mime.subtype()) {
                    (mime::STAR, mime::STAR) => Some(Ordering::Equal),
                    (mime::STAR, _) => Some(Ordering::Less),
                    (_, mime::STAR) => Some(Ordering::Greater),
                    (_, _) => None,
                },
                (mime::STAR, _) => Some(Ordering::Less),
                (_, mime::STAR) => Some(Ordering::Greater),
                (_, _) => None,
            },
            v => v,
        }
    }
}

impl fmt::Display for MediaType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.mime)?;

        if let Some(weight) = self.weight {
            write!(f, ";q={weight}")?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_media_type_should_be_parsed() {
        let t1: MediaType = "text/html; q= 0.5 ".parse().unwrap();
        let t2: MediaType = "application/json".parse().unwrap();
        let t3 = "*/*; q=0.7".parse::<MediaType>().unwrap();
        assert_eq!(t1.mime, Mime::from_str("text/html").unwrap());
        assert_eq!(t1.mime.type_(), mime::TEXT);
        assert_eq!(t1.mime.subtype(), mime::HTML);
        assert_eq!(t2.mime.type_(), mime::APPLICATION);
        assert_eq!(t2.mime.subtype(), mime::JSON);
        assert_eq!(t1.weight, Some(0.5));
        assert_eq!(t2.weight, None);
        assert_eq!(t3.mime.type_(), mime::STAR);
    }

    #[test]
    fn invalid_media_type_should_be_rejected() {
        let t1 = "text/html; q=-0.5".parse::<MediaType>();
        let t2 = "text/html; q=1.5".parse::<MediaType>();
        let t3 = "text/html; q=abcd".parse::<MediaType>();

        assert!(t1.is_err());
        assert_eq!(
            t2.unwrap_err().to_string(),
            "Weight should be 0.0-1.0. Got 1.5"
        );
        assert_eq!(t3.unwrap_err().to_string(), "Invalid weight: abcd");
    }

    #[test]
    fn media_type_should_be_comparable() {
        let t1: MediaType = "text/html; q= 0.5 ".parse().unwrap();
        let t2: MediaType = "application/json".parse().unwrap();
        let t3: MediaType = "text/html".parse().unwrap();
        assert!(t1 < t2);
        assert!(t1 < t3);
        assert!(t2 != t3);
    }

    #[test]
    fn media_type_to_string_should_work() {
        let t1: MediaType = "text/html; q= 0.5 ".parse().unwrap();
        let t2: MediaType = "application/json".parse().unwrap();
        let t3: MediaType = "text/html".parse().unwrap();
        assert_eq!(t1.to_string(), "text/html;q=0.5");
        assert_eq!(t2.to_string(), "application/json");
        assert_eq!(t3.to_string(), "text/html");
    }
}


src/accept.rs
use crate::{error::*, Accept, MediaType};
use http::StatusCode;
use itertools::Itertools;
use mime::Mime;
use std::{cmp::Ordering, fmt, str::FromStr};

impl Accept {
    /// Determine the most suitable `Content-Type` encoding.
    pub fn negotiate(&self, available: &[Mime]) -> Result<Mime, StatusCode> {
        for media_type in &self.types {
            if available.contains(&media_type.mime) {
                return Ok(media_type.mime.clone());
            }
        }

        if self.wildcard.is_some() {
            if available.contains(&mime::TEXT_HTML) {
                return Ok(mime::TEXT_HTML);
            }

            if let Some(accept) = available.iter().next() {
                return Ok(accept.clone());
            }
        }

        Err(StatusCode::NOT_ACCEPTABLE)
    }
}

impl FromStr for Accept {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut types = Vec::new();
        let mut wildcard = None;

        for part in s.split(',') {
            let mtype: MediaType = part.trim().parse()?;

            if mtype.mime.type_() == mime::STAR {
                wildcard = Some(mtype);
            } else {
                types.push(mtype);
            }
        }

        types.sort_by(|a, b| b.partial_cmp(a).unwrap_or(Ordering::Equal));

        Ok(Accept { wildcard, types })
    }
}

impl From<Mime> for Accept {
    fn from(mime: Mime) -> Self {
        Self {
            wildcard: None,
            types: vec![MediaType::from(mime)],
        }
    }
}

impl fmt::Display for Accept {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.types.iter().map(|m| m.to_string()).join(", "))?;

        if let Some(wildcard) = &self.wildcard {
            write!(f, ", {wildcard}")?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accept_should_be_parsed_and_sorted() {
        let accept = "application/json, text/html;q=0.9, text/plain;q=0.8, */*;q=0.7, */*;q=0.6"
            .parse::<Accept>()
            .unwrap();

        assert_eq!(
            accept.wildcard,
            Some(MediaType::from_str("*/*; q=0.6").unwrap())
        );
        assert_eq!(accept.types.len(), 3);
        assert_eq!(
            accept.types[0].mime,
            Mime::from_str("application/json").unwrap()
        );
        assert_eq!(accept.types[0].weight, None);
        assert_eq!(accept.types[1].mime, Mime::from_str("text/html").unwrap());
        assert_eq!(accept.types[1].weight, Some(0.9));
        assert_eq!(accept.types[2].mime, Mime::from_str("text/plain").unwrap());
        assert_eq!(accept.types[2].weight, Some(0.8));
    }

    #[test]
    fn content_negotiation_should_work() {
        let accept = "application/json, text/html;q=0.9, text/plain;q=0.8, */*;q=0.7"
            .parse::<Accept>()
            .unwrap();

        let available = &[
            Mime::from_str("text/html").unwrap(),
            Mime::from_str("application/json").unwrap(),
        ];

        let negotiated = accept.negotiate(&available[..]).unwrap();

        assert_eq!(negotiated, Mime::from_str("application/json").unwrap());

        let available = &[Mime::from_str("application/xml").unwrap()];
        let negotiated = accept.negotiate(&available[..]).unwrap();
        assert_eq!(negotiated, Mime::from_str("application/xml").unwrap());
    }

    #[test]
    fn content_negotiation_should_fail_if_no_available() {
        let accept = "application/json,text/html;q=0.9,text/plain;q=0.8"
            .parse::<Accept>()
            .unwrap();

        let available = &[Mime::from_str("application/xml").unwrap()];
        let negotiated = accept.negotiate(&available[..]);
        assert_eq!(negotiated, Err(StatusCode::NOT_ACCEPTABLE));
    }

    #[test]
    fn accept_to_string_should_work() {
        let accept = "application/json, text/plain;q=0.8, text/html;q=0.9, */*;q=0.7,*/*;q=0.6"
            .parse::<Accept>()
            .unwrap();

        assert_eq!(
            accept.to_string(),
            "application/json, text/html;q=0.9, text/plain;q=0.8, */*;q=0.6"
        );
    }
}
